### ISSUE TITLE
Allow Rename of namespaces to introduce (& maybe remove) namespace parts

### ISSUE THREAD

**Issue by dpoeschl** [dotnet/roslyn#3583]
It would be nice to be able to rename a namespace "NS" to "NS.X" via the rename refactoring.

Removing parts (changing all instances of "NS.X" to "NS") will require more work, both in cases where you have `namespace NS { namespace X { ... }}` and within the Inline Rename experience itself which currently only selects one identifier initially.

This has been requested many times but I couldn't find an issue specifically tracking it.

**Comment by ChristophB125** [comment]
+1 Upvote

**Comment by sandyarmstrong** [comment]
**Version Used**: 3.3

**Steps to Reproduce**:

1. Right-click a namespace in any C# file.
2. Select "Rename" menu item.
3. Type a new namespace that include a "." character.

**Expected Behavior**:

You can type or paste a "." character as long as it is a valid namespace string.

**Actual Behavior**:

You cannot type a "." character. If you attempt to paste one, the outline becomes red, and the popup in the corner of the editor says "The new name is not a valid identifier."

**Bonus**: In an ASP.NET Core app, if you hit return anyway at this point, a yellow bar appears indicating that a `NotSupportedException` has been raised.

```
    <description>System.NotSupportedException: Changing document &apos;C:\Users\sandy\source\repos\WebApplication1\WebApplication1\Pages\_ViewImports.g.cs&apos; is not supported.&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Workspace.CheckAllowedProjectChanges(ProjectChanges projectChanges)&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Workspace.CheckAllowedSolutionChanges(SolutionChanges solutionChanges)&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)&#x000D;&#x000A;   at Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.TryApplyChanges(Solution newSolution, IProgressTracker progressTracker)&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Workspace.TryApplyChanges(Solution newSolution)&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.ApplyRename(Solution newSolution, IWaitContext waitContext)&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.CommitCore(IWaitContext waitContext, Boolean previewChanges)&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.&lt;&gt;c__DisplayClass76_0.&lt;Commit&gt;b__0(IWaitContext waitContext)&#x000D;&#x000A;   at Microsoft.VisualStudio.LanguageServices.Implementation.Utilities.VisualStudioWaitIndicator.Wait(String title, String message, Boolean allowCancel, Boolean showProgress, Action`1 action)&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.InlineRenameSession.Commit(Boolean previewChanges)&#x000D;&#x000A;   at Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.RenameCommandHandler.ExecuteCommand(ReturnKeyCommandArgs args, CommandExecutionContext context)&#x000D;&#x000A;   at Microsoft.VisualStudio.Commanding.CommandHandlerExtensions.ExecuteCommand[T](ICommandHandler commandHandler, T args, Action nextCommandHandler, CommandExecutionContext executionContext)&#x000D;&#x000A;   at Microsoft.VisualStudio.UI.Text.Commanding.Implementation.EditorCommandHandlerService.&lt;&gt;c__DisplayClass13_1`1.&lt;Execute&gt;b__1()&#x000D;&#x000A;   at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.CallExtensionPoint(Object errorSource, Action call, Predicate`1 exceptionFilter)&#x000D;&#x000A;--- End of stack trace from previous location where exception was thrown ---&#x000D;&#x000A;   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)</description>
```

In a simple console app, you can hit return and the namespace rename will be successful.

**Comment by jinujoseph** [comment]
refer https://stackoverflow.com/questions/2871314/change-project-namespace-in-visual-studio

**Comment by dwillis-rpi** [comment]
Hello, just following up here as I opened an [issue](https://developercommunity.visualstudio.com/t/Namespace-Rename-Feature-Does-Not-Allow-/11003244) for this on the Visual Studio forum, which was then redirected to an [issue](https://github.com/dotnet/roslyn/issues/81391) here, which was then closed as a duplicate of this issue.

I'm just curious - how is it possible that this has been open 10 years now and still not fixed? Seems like this should be a very trivial fix? And it stands in the way of people being able to follow Microsoft's own naming guidelines for namespaces...plus, there's no workaround that works anymore (some say that copy/pasting the dot into the "rename" box would work in the past, but that does not work in VS 2022).

I see this thread is tagged with "feature" but it seems more like a bug to me, as the rename refactoring functionality has been in Visual Studio for a long time and this is a piece of it that has never worked correctly.

Is there a blocker that is preventing this from getting done? Maybe the fix is not as trivial as one might assume?

Any insight or update on where this stands in the priority list currently would be appreciated.

Best regards

**Comment by CyrusNajmabadi** [comment]
> how is it possible that this has been open 10 years now and still not fixed?

Other issues have been more important.  In ten years this has only garnered 20 votes.  That's very minor. 

> Seems like this should be a very trivial fix?

It is not. For example. All our code that does things like check for conflicts or verify that the old and new code has the same meaning depends on walking trees if the same structure. Updating that to support different structured trees is non trivial. 

Right now as well we just need to scan out an identifier token. And the entire flow relates to operating on existing tokens and using that that new token. This would involve updating all the code within to now be node based instead of token based. 

Rename is not a string find/replace. It's an actual semantic rewrite. So changing the type of this entity hits a lot of code. 

> I see this thread is tagged with "feature" but it seems more like a bug to me, 

It's a feature request. It's asking for this feature to gain functionality it has never supported.

> as the rename refactoring functionality has been in Visual Studio for a long time 

Length of time isn't relevant. It's still a feature request. 

> and this is a piece of it that has never worked correctly.

The code is working as we intended it to work. If we intended it to support dots and it didn't, then that would be a bug. 

> Is there a blocker that is preventing this from getting done? Maybe the fix is not as trivial as one might assume?

There is no blocker.  But it's not trivial.  And currently it had not been important enough to invest in. 

> Any insight or update on where this stands in the priority list currently would be appreciated.

It is in the backlog. So it is behind all items currently assigned a milestone. 

> Plus, there's no workaround that works anymore (some say that copy/pasting the dot into the "rename" box would work in the past, but that does not work in VS 2022).

If you want a poor man's workaround, rename to `Foo_Bar_Baz`. Then after the rename do a string find/replace it that string and replace with `Foo.Bar.Baz`.

**Comment by dwillis-rpi** [comment]
Hi @CyrusNajmabadi 

Thanks for the response and insight. I can understand most of what you said, except this part:

> It's a feature request. It's asking for this feature to gain functionality it has never supported.

> Length of time isn't relevant. It's still a feature request.

> The code is working as we intended it to work. If we intended it to support dots and it didn't, then that would be a bug.

This doesn't make sense to me. You're saying the fact that it didn't support dots from the beginning was by design? Even though that would essentially make it useless for customers following Microsoft's own [published naming recommendations for namespaces](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-namespaces), since any rename of a namespace with a dot essentially does not work? Seems....strange to say the least - either a design flaw or significant oversight - but if that's the case then so be it I guess.

And if it continues to be treated as a "feature request" then I presume based on your other comments it's unlikely it'll ever get enough momentum to actually happen. This type of stuff is disappointing, as this would otherwise be an extremely useful feature and this single issue stands in the way of that.

I know this thread only has 24 votes but this issue has been mentioned countless times on forums all over the place over many years (as the OP even mentions in their post). The simple fact of the matter is most people probably just continue working around it and considering it a broken feature rather than taking the time to dig through threads to find the one that is the "current" open request for it and vote on it. Which is also what I myself have done for many years.

> If you want a poor man's workaround, rename to Foo_Bar_Baz. Then after the rename do a string find/replace it that string and replace with Foo.Bar.Baz.

Yes, I'm aware I can do a text replacement. The whole point of the rename feature is to be "smarter" and more robust than a simple find/replace. If using find/replace will be the long term recommended "solution" then why even have the rename feature at all?

**Comment by CyrusNajmabadi** [comment]
> This doesn't make sense to me. You're saying the fact that it didn't support dots from the beginning was by design?

Yes. I worked on the original implementation back before Roslyn, and during the rewrite for Roslyn.  It was by design that it did not support dots. 

> Seems....strange to say the least 

It made sense to us. You have an existing name, you are renaming to a new name.  A dotted name is something entirely different and completely upends the implementation and complexity.  

> And if it continues to be treated as a "feature request" then I presume based on your other comments it's unlikely it'll ever get enough momentum to actually happen. 

That isn't the case at all.  


> This type of stuff is disappointing, as this would otherwise be an extremely useful feature and this single issue stands in the way of that.

This is the case with tons of requests on us. 

Note, Roslyn is open source. If you want to add support for this, we could consider supporting it. It's just not something we can justify the investment in over more important issues/features right now. 

> The simple fact of the matter

This is untrue and we see other issues/requests get more votes and attention. So people do seek out and send in feedback. We already connect information from many sources.  We have all sorts of feedback mechanisms and connections to many different communities. These are all considered. 

> Yes, I'm aware I can do a text replacement

I mentioned doing a rename first. So you get those smarts. Then use a find/replace just for the underscore-to-dot part. 

Anyways. It seems that you recognize this is non trivial. I'm not sure what I can tell you at this point. I've explained why it is complex. I've explained why it's low on our priority list. 

If you disagree, you're free to spend effort implementing.  That's the value of open source.  I'm not sure what additional info can be added here.

**Comment by dwillis-rpi** [comment]
@CyrusNajmabadi Fair enough. Some folks will send in feedback if it bothers them enough (sort of like how this eventually bothered me enough to submit a feedback), but at the same time I feel like there are probably a lot more folks who just work around it. In any case, it sounds like it would be a complex undertaking, so I can understand why it would be lower on the priority list with this issue not receiving a ton of votes.

I don't think I'd even try to take a stab at implementing this myself anytime in the near future, as I have -zero- familiarity with this code base and already have my own backlog a mile long. But as you mentioned, this is the value of open sourcing stuff - hopefully either some generous third party with more knowledge than me will make an attempt at this one day, or it will get enough votes to move up enough on the priority list where the main Team decides to implement it.

I do appreciate you taking the time out to reply and provide this context.

Best regards.
